File: src/node/addons/pty.rs
use std::{io::{Read, Write}, sync::mpsc::{self, Receiver, Sender}, time::Duration};

use log::trace;
use portable_pty::{native_pty_system, CommandBuilder, PtySize};
use tao::event_loop::EventLoopProxy;
use tokio::runtime::Runtime;
use wry::RequestAsyncResponder;

use crate::{backend::Backend, common::{respond_404, respond_ok}, node::{common::send_command, node::AppEnv}, types::{BackendCommand, ChildProcess, ElectricoEvents}};

use super::types::PTYCommand;

pub fn process_pty_command(tokio_runtime:&Runtime, _app_env:&AppEnv,
    proxy:EventLoopProxy<ElectricoEvents>,
    backend:&mut Backend,
    command:PTYCommand,
    responder:RequestAsyncResponder,
    _data_blob:Option<Vec<u8>>)  {
    
    let command_sender = backend.command_sender();
    match command {
        PTYCommand::Spawn {id, shell, args, opt } => {
            match native_pty_system().openpty(PtySize {
                rows: opt.rows,
                cols: opt.cols,
                pixel_width: 0,
                pixel_height: 0,
            }) {
                Ok(pair) => {
                    let mut cmd = CommandBuilder::new(shell);
                    cmd.args(args);
                    cmd.cwd(opt.cwd);
                    match pair.slave.spawn_command(cmd) {
                        Ok(_child) => {
                            let (sender, receiver): (Sender<ChildProcess>, Receiver<ChildProcess>) = mpsc::channel();
                            backend.child_process_start(&id, sender.clone());

                            tokio_runtime.spawn(
                                async move {
                                    let mut reader: Box<dyn Read + Send>;
                                    match pair.master.try_clone_reader() {
                                        Ok(r) => {
                                            reader = r;
                                        },
                                        Err(e) => {
                                            log::error!("PTYCommand try_clone_reader Error: {}", e);
                                            return;
                                        }
                                    }
                                    let mut writer: Box<dyn Write + Send>;
                                    match pair.master.take_writer() {
                                        Ok(w) => {
                                            writer = w;
                                        },
                                        Err(e) => {
                                            log::error!("PTYCommand take_writer Error: {}", e);
                                            return;
                                        }
                                    }
                                    let (read_end_sender, read_end_receiver): (Sender<ChildProcess>, Receiver<ChildProcess>) = mpsc::channel();
                                    let write_proxy = proxy.clone();
                                    let write_command_sender = command_sender.clone();
                                    let write_id = id.clone();
                                    tokio::spawn(async move {
                                        trace!("PTYCommand read");
                                        let stdout_buf:&mut [u8] = &mut [0; 65536];
                                        loop {
                                            if let Ok(read) = reader.read(stdout_buf) {
                                                trace!("PTYCommand bytes read {}", read);
                                                if let Ok(d) = read_end_receiver.try_recv() {
                                                    match d {
                                                        ChildProcess::Disconnect => {
                                                            break;
                                                        },
                                                        _ => ()
                                                    }
                                                }
                                                if read>0 {
                                                    let data:Vec<u8> = stdout_buf[0..read].to_vec();
                                                    let _ = send_command(&proxy, &command_sender, BackendCommand::ChildProcessCallback { pid:id.clone(), stream:"stdout".to_string(), data:Some(data) });
                                                }
                                            }
                                        }
                                    });
                                    tokio::spawn(async move {
                                        loop {
                                            if let Ok(cp) = receiver.recv_timeout(Duration::from_millis(100)) {
                                                match cp {
                                                    ChildProcess::StdinWrite { data } => {
                                                        trace!("PTYCommand bytes write {}", data.len());
                                                        let _ = writer.write(data.as_slice());
                                                    },
                                                    ChildProcess::Disconnect => {
                                                        let _ = read_end_sender.send(ChildProcess::Disconnect);
                                                        break;
                                                    },
                                                    _ =>()
                                                }
                                            }
                                        }
                                        trace!("PTYCommand spawn exit");
                                        let _ = send_command(&write_proxy, &write_command_sender, BackendCommand::ChildProcessExit {pid:write_id, exit_code:None});
                                    });
                                }
                            );
                            respond_ok(responder);
                        },
                        Err(e) => {
                            log::error!("PTYCommand spawn_command Error: {}", e);
                            respond_404(responder);
                        }
                    }
                },
                Err(e) => {
                    log::error!("PTYCommand Spawn Error: {}", e);
                    respond_404(responder);
                }
            }
        }
    }
}

File: src/js/backend/addons/pty.js
(function() {
    let uuidv4 = window.__uuidv4;
    let PTY = {
        spawn: (shell, args, opt) => {
            console.log("PTY.spawn", shell, args, opt);
            opt = opt || {};
            args = args || [];
            opt.name = opt.name || "term";
            opt.rows = opt.rows || 30;
            opt.cols = opt.cols || 80;
            opt.cwd = opt.cwd || window.__dirname;
            opt.env = opt.env || {};
            let id = uuidv4();
            let {r, e} = $e_node.syncAddon_PTY_Spawn({"id":id, "shell":shell, "args":args, "opt":opt});
            if (e!=null) {
                throw "PTY.spawn: "+e;
            }
            pcp = {
                on: {}
            }
            window.__electrico.child_process[id] = pcp;
            let ptyProcess = {
                pid:parseInt(r),
                onData: (cb) => {
                    pcp.stdout_on = {
                        data: (data) => {
                            console.log("pty cb out", data.toString());
                            cb({data:data});
                        }
                    }
                },
                write: (data) => {
                    let {r, e} = $e_node.syncChildProcessStdinWrite({pid: id}, data);
                    if (e!=null) {
                        throw "ptyProcess.write error: "+e;
                    }
                },
                resize: (cols, rows) => {
                    // TODO
                    console.error("ptyProcess.resize", cols, rows);
                },
                onExit: (cb) => {
                    pcp.on.close = (exit_code)=> {
                        cb(exit_code);
                    }
                }
            }
            
            return ptyProcess;
        }
    }
    window.__electrico.libs["node-pty"] = PTY;
})();